src/libcore/ptr.rs

```diff
@@ -38,21 +87,65 @@ pub use intrinsics::write_bytes;
 
 /// Executes the destructor (if any) of the pointed-to value.
 ///
-/// This has two use cases:
+/// This is semantically equivalent to calling [`ptr::read`] and discarding
+/// the result, but has the following advantages:
 ///
 /// * It is *required* to use `drop_in_place` to drop unsized types like
 ///   trait objects, because they can't be read out onto the stack and
 ///   dropped normally.
 ///
-/// * It is friendlier to the optimizer to do this over `ptr::read` when
+/// * It is friendlier to the optimizer to do this over [`ptr::read`] when
 ///   dropping manually allocated memory (e.g. when writing Box/Rc/Vec),
 ///   as the compiler doesn't need to prove that it's sound to elide the
 ///   copy.
 ///
+/// [`ptr::read`]: ../ptr/fn.read.html
+///
 /// # Safety
 ///
-/// This has all the same safety problems as `ptr::read` with respect to
-/// invalid pointers, types, and double drops.
+/// Behavior is undefined if any of the following conditions are violated:
+///
+/// * `to_drop` must be [valid] for reads.
+///
+/// * `to_drop` must be properly aligned.
+///
+/// Additionally, if `T` is not [`Copy`], using the pointed-to value after
+/// calling `drop_in_place` can cause undefined behavior. Note that `*to_drop =
+/// foo` counts as a use because it will cause the the value to be dropped
+/// again. [`write`] can be used to overwrite data without causing it to be
+/// dropped.
+///
+/// These restrictions apply even if `T` has size `0`.
+///
+/// [valid]: ../ptr/index.html#safety
+/// [`Copy`]: ../marker/trait.Copy.html
+/// [`write`]: ../ptr/fn.write.html
+///
+/// # Examples
+///
+/// Manually remove the last item from a vector:
+///
+/// ```
+/// use std::ptr;
+/// use std::rc::Rc;
+///
+/// let last = Rc::new(1);
+/// let weak = Rc::downgrade(&last);
+///
+/// let mut v = vec![Rc::new(0), last];
+///
+/// unsafe {
+///     // Without a call `drop_in_place`, the last item would never be dropped,
+///     // and the memory it manages would be leaked.
+///     ptr::drop_in_place(&mut v[1]);
+///     v.set_len(1);
+/// }
+///
+/// assert_eq!(v, &[0.into()]);
+///
+/// // Ensure that the last item was dropped.
+/// assert!(weak.upgrade().is_none());
+/// ```
```

Comment:
I don't fully understand the point that this example is trying to make (I am not that experienced with `Rc` and `Weak`). The following works ([playground](https://play.rust-lang.org/?gist=22e4bc066e8366264b3e7270ace1daf8&version=stable&mode=debug&edition=2015)):

```rust
use std::rc::Rc;
fn main() {
    let last = Rc::new(1);
    let weak = {
        let weak = Rc::downgrade(&last);
        let _ = vec![Rc::new(0), last];
        weak
    };
    assert!(weak.upgrade().is_none());
}
```

so I don't see why the call to `drop_in_place` would be necessary to avoid the memory leak. What am I missing?

Reply:
The key thing here is the `v.set_len(1)` which essentially `mem::forget`s the last element of the `Vec`.

Reply:
Ah! That makes sense now! When I read this, I though that one had to do both the `drop_in_place` and `set_len` for the last item to always be dropped, and I was thinking that if I don't do any of those, it is dropped as well.